osu.Game/Skinning/LegacyTextureLoaderStore.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.IO.Stores;
+using SixLabors.ImageSharp;
+using SixLabors.ImageSharp.Processing;
+
+namespace osu.Game.Skinning
+{
+    public class LegacyTextureLoaderStore : IResourceStore<TextureUpload>
+    {
+        private readonly IResourceStore<TextureUpload>? wrappedStore;
+
+        public LegacyTextureLoaderStore(IResourceStore<TextureUpload>? wrappedStore)
+        {
+            this.wrappedStore = wrappedStore;
+        }
+
+        public TextureUpload Get(string name)
+        {
+            var textureUpload = wrappedStore?.Get(name);
+
+            if (textureUpload == null)
+                return null!;
+
+            return shouldConvertToGrayscale(name)
+                ? convertToGrayscale(textureUpload)
+                : textureUpload;
+        }
+
+        public Task<TextureUpload> GetAsync(string name, CancellationToken cancellationToken = new CancellationToken())
+        {
+            var textureUpload = wrappedStore?.Get(name);
+
+            if (textureUpload == null)
+                return null!;
+
+            return shouldConvertToGrayscale(name)
+                ? Task.Run(() => convertToGrayscale(textureUpload), cancellationToken)
+                : Task.FromResult(textureUpload);
+        }
+
+        // https://github.com/peppy/osu-stable-reference/blob/013c3010a9d495e3471a9c59518de17006f9ad89/osu!/Graphics/Textures/TextureManager.cs#L91-L96
+        private static readonly string[] grayscale_sprites =
+        {
+            @"taiko-bar-right",
+            @"taikobigcircle",
+            @"taikohitcircle",
+            @"taikohitcircleoverlay"
+        };
+
+        private bool shouldConvertToGrayscale(string name)
+        {
+            foreach (string grayscaleSprite in grayscale_sprites)
+            {
+                // unfortunately at this level of lookup we can encounter `@2x` scale suffixes in the name,
+                // so straight equality cannot be used.
+                if (name.StartsWith(grayscaleSprite, StringComparison.OrdinalIgnoreCase))
```

Comment:
I think this has to be an exact match otherwise it breaks e.g. kiai glow (`taiko-bar-right-glow`):

<img width="1624" alt="image" src="https://github.com/ppy/osu/assets/1329837/43a09bac-9b2d-4d1f-838e-a1d8e9f5f315">

Reply:
ouch. good point.

i think https://github.com/ppy/osu/pull/27059/commits/c21af1bf3d44f3c5b70c95c573748c2f41cd35f1 should work....?